---
title: "Plots for article"
output: html_notebook
---


```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
knitr::opts_chunk$set(echo=FALSE, message = FALSE)
```

```{r, results = 'hide'}
library(tidyverse)
library(Seurat)
library(ggpubr)
library(venn)
library(clusterProfiler)
library(biomaRt)
```

```{r}
load(file = "data/rna.integrated.filtered.after.clustering.obj.RData")
rna.integrated.cluster.blood <- subset(rna.integrated, idents = 0)
rna.integrated.cluster.rest <- subset(rna.integrated, idents = c(1,2))
```


```{r}
rna_qc_all_plot <- VlnPlot(rna.integrated, 
        features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        ncol = 3, 
        pt.size = 0.1, 
        group.by = "orig.ident", 
        assay = "RNA") 
png(filename = "plots/for_article_png/Suppl_scRNA_all_samples_QC.png", 
    res = 400, 
    units = "in",
    width = 13,
    height = 11)
print(rna_qc_all_plot)
dev.off()
png(filename = "plots/for_article/Suppl_scRNA_all_samples_QC.png", 
    res = 400, 
    units = "in",
    width = 13,
    height = 11)
print(rna_qc_all_plot)
dev.off()
rna_qc_all_plot
```


```{r}
png(filename = "plots/for_article/S_f_1_umap_all_by_samples.png", res = 300, width = 1800, height =900)
DimPlot(rna.integrated, reduction = "umap", group.by = "orig.ident")
dev.off()
```

```{r}
png(filename = "plots/for_article/f_1_umap_all_low_res.png", res = 300, width = 1500, height =1200)
DimPlot(rna.integrated, reduction = "umap", cols = c("#EF9A9A", "#80CBC4", "#9FA8DA"))
dev.off()
DimPlot(rna.integrated, reduction = "umap", cols = c("#EF9A9A", "#80CBC4", "#9FA8DA"))
```

```{r}
paste0("Number of cells in the whole data: ", dim(rna.integrated@assays[["RNA"]]@counts)[[2]])
```

```{r}
rna.integrated.cluster.rest <- FindNeighbors(rna.integrated.cluster.rest, dims = 1:15)
rna.integrated.cluster.rest  <- FindClusters(rna.integrated.cluster.rest, resolution = 0.3)
rna.integrated.cluster.rest <- RunUMAP(rna.integrated.cluster.rest, dims = 1:15)

# rna.integrated.cluster.rest <- FindClusters(rna.integrated.cluster.rest, resolution = 0.1)
```

```{r, fig.width = 13, fig.height = 13}
rna.integrated.cluster.rest  <- FindClusters(rna.integrated.cluster.rest, resolution = 0.05)
umap_005_res_plot <- DimPlot(rna.integrated.cluster.rest, reduction = "umap", pt.size = 1.5) + 
  theme(text=element_text(size=20),
        axis.text=element_text(size=20),
        legend.text=element_text(size=25)) +
  labs(title = "UMAP plot with resolution = 0.05")

rna.integrated.cluster.rest  <- FindClusters(rna.integrated.cluster.rest, resolution = 0.1)
umap_01_res_plot <- DimPlot(rna.integrated.cluster.rest, reduction = "umap", pt.size = 1.5) + 
  theme(text=element_text(size=20),
        axis.text=element_text(size=20),
        legend.text=element_text(size=25)) +
    labs(title = "UMAP plot with resolution = 0.1")

rna.integrated.cluster.rest  <- FindClusters(rna.integrated.cluster.rest, resolution = 0.3)
umap_03_res_plot <- DimPlot(rna.integrated.cluster.rest, reduction = "umap", pt.size = 1.5) + 
  theme(text=element_text(size=20),
        axis.text=element_text(size=20),
        legend.text=element_text(size=25)) +
    labs(title = "UMAP plot with resolution = 0.3")

rna.integrated.cluster.rest  <- FindClusters(rna.integrated.cluster.rest, resolution = 1.5)
umap_15_res_plot <- DimPlot(rna.integrated.cluster.rest, reduction = "umap", pt.size = 1.5) + 
  theme(text=element_text(size=20),
        axis.text=element_text(size=20),
        legend.text=element_text(size=25)) +
    labs(title = "UMAP plot with resolution = 1.5")

rna.integrated.cluster.rest  <- FindClusters(rna.integrated.cluster.rest, resolution = 0.3)

png(file = "plots/for_article/Suppl_scRNA_umap_rest_testing_different_resolutions.png", 
    res = 300, 
    units = "in",
    width = 13,
    height = 13)
(umap_005_res_plot | umap_01_res_plot) / (umap_03_res_plot | umap_15_res_plot)
dev.off()

(umap_005_res_plot | umap_01_res_plot) / (umap_03_res_plot | umap_15_res_plot)
```

```{r}
# Separating neuron cluster
# integrated_rest_extreme_res  <- FindClusters(rna.integrated.cluster.rest, resolution = 6)
standalone_cluster_cell_IDs <- c( "cDNA1_ATCCTATCAAAGGCTG-1", "cDNA1_GACTGATGTACGGTTT-1", "cDNA1_GGTTGTACAAGACTGG-1", "cDNA1_TTCTAGTAGTCACTGT-1", "cDNA2_AATAGAGGTGGAAGTC-1", "cDNA2_ACTATCTTCAGGAAAT-1", "cDNA2_CCACGAGCACATTCGA-1", "cDNA2_CGTTCTGCAGTTCCAA-1", "cDNA2_CTCGAGGTCCACTTCG-1", "cDNA2_GAACGTTAGAGAACCC-1", "cDNA2_GCGGAAAAGACTGTTC-1", "cDNA2_GGAATGGAGGTGCTAG-1", "cDNA2_GGTAGAGGTGTTGAGG-1", "cDNA2_GTCATCCGTGAGCGAT-1", "cDNA2_GTGAGTTCAGCTCTGG-1", "cDNA2_TCTCAGCTCCGGACGT-1")
new_cluster_ids <- as.character(Idents(rna.integrated.cluster.rest))
names(new_cluster_ids) <- names(Idents(rna.integrated.cluster.rest))
new_cluster_ids <- replace(new_cluster_ids,standalone_cluster_cell_IDs,"neurons")
new_cluster_ids <- as.factor(new_cluster_ids)
# integrated_rest_extreme_res <- StashIdent(object = integrated_rest_extreme_res, save.name = "auto.ident")
Idents(object = rna.integrated.cluster.rest) <- new_cluster_ids
```


```{r}
new_cluster_labels <- c("leukocytes_1", "macrophages", "gill_integument", "low_quality", "neutrophils", "lymphocytes", "ionocytes", "epithelium", "leukocytes_2", "neurons")
names(new_cluster_labels) <- levels(rna.integrated.cluster.rest)
rna.integrated.cluster.rest_annotated <- RenameIdents(rna.integrated.cluster.rest, new_cluster_labels)
```

```{r}
saveRDS(rna.integrated.cluster.rest_annotated, file = "data/rna_integrated_cluster_rest_annotated.Rds")
```


```{r}
#Set color pallete for Clusters
nclusters <- length(levels(Idents(rna.integrated.cluster.rest_annotated)))
cluster_color_pallete <- RColorBrewer::brewer.pal(n = nclusters, name = 'Paired')
```



```{r, fig.width = 10, fig.height = 7}
pdf(file = "plots/for_article/S_f_2_umap_rest_high_res_03_annotated.pdf", width = 10, height = 7)
DimPlot(rna.integrated.cluster.rest_annotated, reduction = "umap", cols = cluster_color_pallete, pt.size = 1.5) + 
  theme(text=element_text(size=20),
        axis.text=element_text(size=20),
        legend.text=element_text(size=25))
dev.off()
DimPlot(rna.integrated.cluster.rest_annotated, reduction = "umap", cols = cluster_color_pallete, pt.size = 1.5) + 
  theme(text=element_text(size=20),
        axis.text=element_text(size=20),
        legend.text=element_text(size=25))
```


```{r, fig.width = 10, fig.height = 7}
pdf(file = "plots/for_article/Suppl_scRNA_umap_rest_high_res_03_annotated_with_lables.pdf", width = 10, height = 7)
DimPlot(rna.integrated.cluster.rest_annotated, reduction = "umap", label = TRUE, repel = TRUE, cols = cluster_color_pallete, label.size = 7, pt.size = 1.5) + 
  theme(text=element_text(size=20),
        axis.text=element_text(size=20),
        legend.text=element_text(size=25))
dev.off()
DimPlot(rna.integrated.cluster.rest_annotated, reduction = "umap", label = TRUE, repel = TRUE, cols = cluster_color_pallete, label.size = 7, pt.size = 1.5)  + 
  theme(text=element_text(size=20),
        axis.text=element_text(size=20),
        legend.text=element_text(size=25))
```


```{r}
paste0("Number of cells in the non-blood data: ", dim(rna.integrated.cluster.rest_annotated@assays[["RNA"]]@counts)[[2]])
```

### Heatmap rest

```{r fig.height=42, fig.width=9}
integrated.cluster.rest.markers <- FindAllMarkers(rna.integrated.cluster.rest_annotated, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.15)
integrated.cluster.rest.markers %>%
    group_by(cluster) %>%
    top_n(n = 20, wt = avg_log2FC) -> top_by_logfc

integrated.cluster.rest.markers_RNA <- FindAllMarkers(rna.integrated.cluster.rest_annotated, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.15, assay = "RNA")
integrated.cluster.rest.markers_RNA %>%
    group_by(cluster) %>%
    # filter(avg_log2FC > 1) %>% 
    top_n(n = 20, wt = avg_log2FC) -> top_on_RNA

FindAllMarkers(rna.integrated.cluster.rest_annotated, 
               test.use = "MAST",
               only.pos = TRUE, 
               min.pct = 0.1, 
               logfc.threshold = 0.15, 
               assay = "RNA"
               ) %>%
    group_by(cluster) %>%
    top_n(n = 20, wt = avg_log2FC) -> top_MAST

manual_markers <- tibble(p_val = NA,
                        avg_log2FC = NA,
                        pct.1 = NA,
                        pct.2 = NA,
                        p_val_adj = NA,
                        cluster = factor(c("lymphocytes", "macrophages"), levels = levels(Idents(rna.integrated.cluster.rest_annotated))),
                        gene = c("lck", "mpeg1.2")
                        )

exclude_markers <- c("atp6v1g1", "UBL5", "ENSGACG00000009225", "acod1", "tsr3", "atp6v1e1b", "rpl22l1", "ENSGACG00000020069", "ddx18", "cnbpa", "sh2d1ab", "ENSGACG00000012769", "tnfsf14", "ENSGACG00000004750", "ENSGACG00000009171", "ENSGACG00000018643", "sla1a", "tagln2", "rpl19", "sftpbb", "nme2a", "nkl.4", "rpl29", "eif3i", "zgc:162944", "tubb2b", "ran", "rps13", "rps7", "fance", "rpl23a", "ENSGACG00000000326", "rplp2l", "ENSGACG00000000860", "ENSGACG00000007836", "rplp0", "rps26", "RPL19 (1 of many)", "rpl13a", "atp6v1g1", "UBL5", "ENSGACG00000009225", "ENSGACG00000007536", "ENSGACG00000001108", "cndp2", "fabp3", "zgc:152830", "mdh1aa", "ENSGACG00000000869", "rpl35a", "rps9", "rpl35", "ENSGACG00000013895", "HBE1 (1 of many)", "ENSGACG0000001390", "ENSGACG00000001394", "hbae5", "h2afva", "scinla", "ENSGACG00000020692", "ENSGACG00000001436", "capn2b", "chmp4c", "chmp2a", "gstr (1 of many)", "rhbg", "rdh12l", "si:dkey−87o1.2", "ENSGACG00000008205", "id4", "krt8", "ENSGACG00000006166", "ENSGACG00000006236", "ENSGACG00000004882", "ENSGACG00000017697", "nrarpa", "capn2b", "zgc:86896", "camk2g2", "cyp46a1.4", "plcg2", "plecb", "ENSGACG00000013902", "tpm4a", "sptbn2", "lmo7a", "pfkpb", "CKB", "itgb2", "tank", "scinlb", "nop56", "ENSGACG00000007176", "ENSGACG00000006275", "sat1b", "si:dkey−87o1.2", "HBE1 (1 of many).1")

rbind(top_by_logfc, top_on_RNA, top_MAST, manual_markers) %>% 
  transform(gene_cluster = paste0(gene, "_", cluster)) %>% 
  group_by(gene_cluster) %>%
  sample_n(1) %>%  # dropping duplicate markers
  arrange(cluster, -avg_log2FC) %>% 
  ungroup() %>% 
  dplyr::select(-c(gene_cluster)) %>% 
  write_delim(file = "data/Suppl_table_scRNA_no_erythro_cluster_markers_res_03_labeled.tsv", delim = "\t")

rbind(top_by_logfc, top_on_RNA, top_MAST, manual_markers) %>% 
  transform(gene_cluster = paste0(gene, "_", cluster)) %>% 
  group_by(gene_cluster) %>%
  sample_n(1) %>%  # dropping duplicate markers
  arrange(cluster, -avg_log2FC) %>% 
  filter((gene %in% exclude_markers) == FALSE) -> top_markers_comb
# Number of cell with non-zero expression per cluster 
#     convert all non-zero values to 1
rest_count_matrix_binarized <- rna.integrated.cluster.rest_annotated@assays[["RNA"]]@counts # all non-zero peaks set to value 1
rest_count_matrix_binarized@x <- as.numeric(rep_len(1L, length(rest_count_matrix_binarized@x)))
#     sum column-wise by gene (resulting with a matrix of num_nonzero_peaks_in_gene vs cells)
expressing_cells_per_cluster_matrix <- t(rowsum(t(as.matrix(rest_count_matrix_binarized)), Idents(rna.integrated.cluster.rest_annotated)))

#     normalizing for the cluster size
norm_vector <- as.vector(table(Idents(rna.integrated.cluster.rest_annotated))/100)
expressing_cells_per_cluster_matrix <- t(t(expressing_cells_per_cluster_matrix)/norm_vector)

to_heatmap_rest <- expressing_cells_per_cluster_matrix[top_markers_comb$gene,]
rows_annot <- as.data.frame(top_markers_comb[,c("gene_cluster", "cluster")])
rownames(rows_annot)<- make.unique(rows_annot$gene_cluster)
rows_annot$gene_cluster <- NULL


annot_colors <- cluster_color_pallete
names(annot_colors) <- levels(Idents(rna.integrated.cluster.rest_annotated))
annot_colors <- list(cluster = annot_colors)

gene_names_heatmap <- rownames(to_heatmap_rest)
rownames(to_heatmap_rest) <- top_markers_comb$gene_cluster

breakpoints <- seq(0, 100, by = 10)
breakpoints[2] <- 0.1
cols <- colorRampPalette(RColorBrewer::brewer.pal(n = 11, name = "PuBu"))(length(breakpoints))


pheatmap::pheatmap(t(scale(t(to_heatmap_rest), center = FALSE)), 
                   cluster_cols = FALSE, 
                   cluster_rows = FALSE,  
                   labels_row = gene_names_heatmap,
                   # breaks = breakpoints,
                   # color = cols,
                   display_numbers = round(to_heatmap_rest, 1), # Side loading different values to display ! https://www.biostars.org/p/317989/
                   annotation_colors = annot_colors,
                   # scale = "row",
                   annotation_row = rows_annot
                   )-> heatmap_rest_avg_with_num
pdf(file = "plots/for_article/Suppl_scRNA_non_erythrocytes_heatmap_markers_10_clusters_percentage_of_expressing_cells_in_a_cluster_MORE_MARKERS_with_numbers.pdf", height = 42)
heatmap_rest_avg_with_num
dev.off()

png(filename = "plots/for_article/Suppl_scRNA_non_erythrocytes_heatmap_markers_10_clusters_percentage_of_expressing_cells_in_a_cluster_MORE_MARKERS_with_numbers.png", height = 42, units = "in", res = 300, width = 7)
heatmap_rest_avg_with_num
dev.off()
```


```{r}
best_markers <- c("ENSGACG00000001123", "arpc1b", "coro1a", "ENSGACG00000014576", "ENSGACG00000009315", "swap70a", "thrap3a", "mpeg1.2", "ENSGACG00000018764", "krt99", "ca15b", "TYMP", "alas2", "tmem205", "gp1bb", "slc4a1a", "npsn (1 of many)", "fbp2", "ncf2", "mpx", "lck", "ENSGACG00000004719", "prf1.9", "ENSGACG00000006316", "atp1a1a.4", "slc4a1b", "atp1b1b", "cyc1", "krt15", "S100P", "ptgdsb.1", "ENSGACG00000011970", "ctsa", "lgmn", "ENSGACG00000003551", "cxcl19", "s100a1", "smpx", "camsap3", "tspan2a")
best_markers <- top_markers_comb[top_markers_comb$gene %in% best_markers,]

# Number of cell with non-zero expression per cluster 
#     convert all non-zero values to 1
rest_count_matrix_binarized <- rna.integrated.cluster.rest_annotated@assays[["RNA"]]@counts # all non-zero peaks set to value 1
rest_count_matrix_binarized@x <- as.numeric(rep_len(1L, length(rest_count_matrix_binarized@x)))
#     sum column-wise by gene (resulting with a matrix of num_nonzero_peaks_in_gene vs cells)
expressing_cells_per_cluster_matrix <- t(rowsum(t(as.matrix(rest_count_matrix_binarized)), Idents(rna.integrated.cluster.rest_annotated)))

#     normalizing for the cluster size
norm_vector <- as.vector(table(Idents(rna.integrated.cluster.rest_annotated))/100)
expressing_cells_per_cluster_matrix <- t(t(expressing_cells_per_cluster_matrix)/norm_vector)

to_heatmap_rest <- expressing_cells_per_cluster_matrix[best_markers$gene,]
rows_annot <- as.data.frame(best_markers[,c("gene_cluster", "cluster")])
rownames(rows_annot)<- make.unique(rows_annot$gene_cluster)
rows_annot$gene_cluster <- NULL


annot_colors <- cluster_color_pallete
names(annot_colors) <- levels(Idents(rna.integrated.cluster.rest_annotated))
annot_colors <- list(cluster = annot_colors)

gene_names_heatmap <- rownames(to_heatmap_rest)
rownames(to_heatmap_rest) <- best_markers$gene_cluster

pheatmap::pheatmap(t(scale(t(to_heatmap_rest), center = FALSE)), 
                   cluster_cols = FALSE, 
                   cluster_rows = FALSE,  
                   labels_row = gene_names_heatmap,
                   # breaks = breakpoints,
                   # color = cols,
                   # display_numbers = round(to_heatmap_rest, 1), # Side loading different values to display ! https://www.biostars.org/p/317989/
                   annotation_colors = annot_colors,
                   # scale = "row",
                   annotation_row = rows_annot
                   )-> heatmap_rest_avg_with_num
pdf(file = "plots/for_article/scRNA_non_erythrocytes_heatmap_10_clusters_percentage_of_expressing_cells_in_a_cluster_BEST_MARKERS.pdf", height = 7, width = 5)
heatmap_rest_avg_with_num
dev.off()
```

### Average expression of markers on UMAP plot

```{r fig.width=9, fig.height=9}
DefaultAssay(rna.integrated.cluster.rest_annotated) <- "RNA"
avg_plot_macrophages <- Nebulosa::plot_density(rna.integrated.cluster.rest_annotated, c("thrap3a"), slot = "counts", )
avg_plot_integument <- Nebulosa::plot_density(rna.integrated.cluster.rest_annotated, c("ca15b"), slot = "counts")
avg_plot_neutrophils <- Nebulosa::plot_density(rna.integrated.cluster.rest_annotated, c("npsn (1 of many)"), slot = "counts")
avg_plot_lymphocytes <- Nebulosa::plot_density(rna.integrated.cluster.rest_annotated, c("prf1.9"), slot = "counts")
avg_plot_ionocytes <- Nebulosa::plot_density(rna.integrated.cluster.rest_annotated, c("atp1b1b"), slot = "counts")
avg_plot_epithelium <- Nebulosa::plot_density(rna.integrated.cluster.rest_annotated, c("ptgdsb.1"), slot = "counts")
avg_plot_leukocytes_2 <- Nebulosa::plot_density(rna.integrated.cluster.rest_annotated, c("cxcl19"), slot = "counts")
avg_plot_neurons <- Nebulosa::plot_density(rna.integrated.cluster.rest_annotated, c("s100a1"), slot = "counts")

full_plot <- ((avg_plot_macrophages + ggtitle("macrophages")) + 
    (avg_plot_integument+ ggtitle("integument")) + 
    (avg_plot_neutrophils) + ggtitle("neutropils")) / 
  ((avg_plot_lymphocytes + ggtitle("lymphocytes")) + 
     (avg_plot_ionocytes + ggtitle("ionocytes")) + 
     (avg_plot_epithelium + ggtitle("epithelium"))) / 
((avg_plot_leukocytes_2 + ggtitle("leukocytes_2")) + 
   (avg_plot_neurons + ggtitle("neurons")) + 
   patchwork::plot_spacer()) & 
  theme(
    legend.position = "none",
        axis.title = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank(),
        axis.line = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_blank())


pdf(file = "plots/for_article/scRNA_avg_marker_expression_on_UMAP.pdf", width = 9, height = 9)
full_plot
dev.off()

DefaultAssay(rna.integrated.cluster.rest_annotated) <- "integrated"
```



### Fig. 1d. Correlation of marine-freshwater expression divergence between bulk RNA-seq and averaged snRNA-seq datasets. 

```{r prepping_sc_DEGs}
# Loading DEGs generated by Seurat for the whole dataset saline vs freshwater
sc_DEGs <- read.csv(file = "data/DEGs.by.salinity.all.csv", header = TRUE)
id_match_table <- read.table(file = "data/10X_features.tsv", header = FALSE, sep = "\t", col.names = c("ensembl", "mixed", "type"))[,1:2]
sc_DEGs <- merge(sc_DEGs, id_match_table, by.x = "X", by.y = "mixed")
rownames(sc_DEGs) <- sc_DEGs$ensembl

sc_DEGs_filtered <- subset(sc_DEGs, abs(avg_logFC) > 0.25)
```

```{r calculation_bulk_DEGs}
bulk_DEGs <- as.data.frame(readxl::read_xlsx("data/S5C_table_RNAseq_DEG_MMFF.xlsx"))
rownames(bulk_DEGs) <- bulk_DEGs$...1
bulk_DEGs_common <- bulk_DEGs[sc_DEGs$ensembl, ]

bulk_DEGs_common_filtered <- bulk_DEGs[sc_DEGs_filtered$ensembl, ]
```

```{r plotting_corr_bulk_vs_sc}

#combining DEG tablesf
combined_DEGs <- merge(sc_DEGs[,c("avg_logFC", "p_val", "ensembl")], bulk_DEGs_common[,c("logFC", "PValue", "...1")], by.x = "ensembl", by.y = "...1")
colnames(combined_DEGs) <- c("ensembl", "single_cell_logFC", "single_cell_pval", "bulk_logFC", "bulk_pval")
rownames(combined_DEGs) <- combined_DEGs$ensembl
write.table(combined_DEGs, file = "data/DEGs_overlapping_sc_and_bulk.tsv", sep = "\t")
combined_DEGs$density <- densCols(combined_DEGs$single_cell_logFC, combined_DEGs$bulk_logFC, colramp = colorRampPalette(rev(rainbow(10, end = 4/6))), nbin = 1024)



p_corr_bulk_vs_sc <- ggplot(combined_DEGs) +
      geom_point(aes(single_cell_logFC, bulk_logFC, col = density), size = 1) +
      stat_cor(aes(single_cell_logFC, bulk_logFC), method = "pearson", label.x = 0.03, label.y = -2, size=6, color="#696969")+
      geom_smooth(aes(single_cell_logFC, bulk_logFC), method=lm, se=FALSE, fullrange=TRUE, linetype="dashed", color="dark grey") +
      coord_cartesian(xlim = c(-0.15, 0.15), ylim = c(-2.5, 2.25)) +
      # xlim(-0.25, 0.25) +
      # ylim(-2.5, 2.25) +
      scale_color_identity() +
      theme_classic(base_size = 20) +
      theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) +
      xlab("logFC, scRNA-seq") +
      ylab("logFC, RNA-seq")

p_corr_bulk_vs_sc
svg(filename = "plots/for_article/f_1d_scatter_scRNA_vs_bulk_with_density.svg", width = 7, height = 7)
print(p_corr_bulk_vs_sc)
dev.off()
```



Insignificant, should not be included in the paper.

```{r number_of_cells_in_data}
# paste0("Number of cells in the whole dataset: ", dim(rna.integrated.cluster.rest@assays[["RNA"]]@counts)[[2]],
       # "\n Percent ")
```

### F-test histogram for blood cluster
```{r}
f.test.filtered.blood <- read.table(file = "data/f.test.filtered.blood.tsv", sep = "\t", header = TRUE)
p_f_ratio_blood <- ggplot(f.test.filtered.blood, aes(x=f_ratio))+
  geom_histogram(binwidth = 0.25) +
  geom_vline(aes(xintercept=median(f_ratio)), color="blue", linetype="dashed", size=1) +
  coord_cartesian(xlim = c(0, 5)) +
  theme_classic(base_size = 20) +
  xlab("F-ratio") +
  ylab("Number of genes")

svg(filename = "plots/for_article/f_1f_F_ratio_histogram_for_blood_cluster.svg")
print(p_f_ratio_blood)
dev.off()
p_f_ratio_blood
```


```{r dist_hist_blood_cluster_btw_phenotype_variability}
load(file = "data/rna.integrated.filtered.after.clustering.obj.RData")
rna.integrated.cluster.blood <- subset(rna.integrated, idents = 0)

rna.integrated.cluster.blood <- FindNeighbors(rna.integrated.cluster.blood, dims = 1:15)
rna.integrated.cluster.blood  <- FindClusters(rna.integrated.cluster.blood, resolution = 0.3)
rna.integrated.cluster.blood <- RunPCA(rna.integrated.cluster.blood)

set.seed(42)
pca_raw_blood <- as.data.frame(rna.integrated.cluster.blood@reductions[["pca"]]@cell.embeddings)
pca_raw_blood$sample <- rna.integrated.cluster.blood@meta.data[["orig.ident"]]
pca_raw_blood <- pca_raw_blood[sample(nrow(pca_raw_blood)), c("PC_1", "PC_2","PC_3","PC_4","PC_5","PC_6","PC_7","PC_8","PC_9","PC_10","sample")]

pca_saline_1_half <- subset(pca_raw_blood, sample == "stickleback.sample.1")[,1:10]
pca_saline_2_half <- subset(pca_raw_blood, sample == "stickleback.sample.2")[,1:10]
dist_matrix_salineVsaline <- fields::rdist(pca_saline_1_half, pca_saline_2_half)

pca_freshwater_1_half <- subset(pca_raw_blood, sample == "stickleback.sample.3")[,1:10]
pca_freshwater_2_half <- subset(pca_raw_blood, sample == "stickleback.sample.4")[,1:10]

dist_matrix_freshwaterVfreshwater <- fields::rdist(pca_freshwater_1_half, pca_freshwater_2_half)

frame_for_hist <- data.frame(dist = c(as.vector(dist_matrix_salineVsaline), as.vector(dist_matrix_freshwaterVfreshwater)),
                             pair = c(rep("salineVsaline", length(dist_matrix_salineVsaline)),
                                      rep("freshwaterVfreshwater", length(dist_matrix_freshwaterVfreshwater))))

mu_btw <- plyr::ddply(frame_for_hist, "pair", summarise, grp.mean=mean(dist))

small_frame_for_hist <- frame_for_hist[sample(nrow(frame_for_hist), 100000),]

p <- ggplot(small_frame_for_hist, aes(x=dist, fill=pair, color=pair)) +
  geom_histogram(aes(y = ..ncount..), position="identity", alpha=0.5, binwidth = 0.15) + 
  geom_vline(data=mu_btw, aes(xintercept=grp.mean, color=pair),
             linetype="dashed")+
  coord_cartesian(xlim = c(0, 15)) +
  theme_classic(base_size = 20) +
  xlab("Distance between pair of cells") +
  ylab("Number cell pairs")
p
svg(filename = "plots/for_article/f_1e_histogram_of_distance_btw_cells_within_saline_and_freshwater_only_blood_equalized_heights.svg", width = 11, height = 6)
print(p)
dev.off()
```


```{r pairwise_histogram_with_downsampling, fig.width=11}
load(file = "data/rna.integrated.filtered.after.clustering.obj.RData")
rna.integrated.cluster.blood <- subset(rna.integrated, idents = 0)

rna.integrated.cluster.blood <- FindNeighbors(rna.integrated.cluster.blood, dims = 1:15)
rna.integrated.cluster.blood  <- FindClusters(rna.integrated.cluster.blood, resolution = 0.3)
rna.integrated.cluster.blood <- RunPCA(rna.integrated.cluster.blood)

set.seed(42)
pca_raw_blood <- as.data.frame(rna.integrated.cluster.blood@reductions[["pca"]]@cell.embeddings)
pca_raw_blood$sample <- rna.integrated.cluster.blood@meta.data[["orig.ident"]]
pca_raw_blood <- pca_raw_blood[sample(nrow(pca_raw_blood)), c("PC_1", "PC_2","PC_3","PC_4","PC_5","PC_6","PC_7","PC_8","PC_9","PC_10","sample")]

pca_saline_1_half <- subset(pca_raw_blood, sample == "stickleback.sample.1")[,1:10]
pca_saline_2_half <- subset(pca_raw_blood, sample == "stickleback.sample.2")[,1:10]
pca_saline_2_half <- pca_saline_2_half[sample(nrow(pca_saline_2_half), nrow(pca_saline_1_half)),] # ========== DOWNSAMPLING CELL COUNT TO BALANCE SAMPLING FROM BOTH FISH
 
dist_matrix_salineVsaline <- fields::rdist(pca_saline_1_half, pca_saline_2_half)

pca_freshwater_1_half <- subset(pca_raw_blood, sample == "stickleback.sample.3")[,1:10]
pca_freshwater_1_half <- pca_freshwater_1_half[sample(nrow(pca_freshwater_1_half), nrow(pca_saline_1_half)),]
pca_freshwater_2_half <- subset(pca_raw_blood, sample == "stickleback.sample.4")[,1:10]
pca_freshwater_2_half <- pca_freshwater_2_half[sample(nrow(pca_freshwater_2_half), nrow(pca_freshwater_1_half)),]

dist_matrix_freshwaterVfreshwater <- fields::rdist(pca_freshwater_1_half, pca_freshwater_2_half)

frame_for_hist <- data.frame(dist = c(as.vector(dist_matrix_salineVsaline), as.vector(dist_matrix_freshwaterVfreshwater)),
                             pair = c(rep("salineVsaline", length(dist_matrix_salineVsaline)),
                                      rep("freshwaterVfreshwater", length(dist_matrix_freshwaterVfreshwater))))

mu_btw <- plyr::ddply(frame_for_hist, "pair", summarise, grp.mean=mean(dist))

small_frame_for_hist <- frame_for_hist[sample(nrow(frame_for_hist), 100000),]

p <- ggplot(small_frame_for_hist, aes(x=dist, fill=pair, color=pair)) +
  geom_histogram(aes(y = ..ncount..), position="identity", alpha=0.5, binwidth = 0.15) + 
  geom_vline(data=mu_btw, aes(xintercept=grp.mean, color=pair),
             linetype="dashed")+
  # coord_cartesian(xlim = c(0, 15)) +
  theme_classic(base_size = 20) +
  xlab("Distance between pair of cells") +
  ylab("Number cell pairs")
p
svg(filename = "plots/for_article/Suppl_histogram_of_distance_btw_cells_within_saline_and_freshwater_only_blood_equalized_heights_WITH_DOWNSAMPLING.svg", width = 11, height = 6)
print(p)
dev.off()
```

```{r means_table_blood_cluster_btw_phenotype_variability}
mu_btw
```

# F-ratio without filtering for erytrocyte cluster for selected HVGs

```{r subsetting_blood_for_each_sample_and_HVGs}
load(file = "data/rna.integrated.filtered.after.clustering.obj.RData")
rna.integrated.cluster.blood <- subset(rna.integrated, idents = 0)
blood_cells_ids <- rna.integrated.cluster.blood@assays[["RNA"]]@counts@Dimnames[[2]]
# blood_cells_ids <- paste(blood_cells_ids, rep("-1", length(blood_cells_ids)), sep = "") 


stickleback.rna.comb <- readRDS(file = "data/stickleback.rna.seurat.samples.PURE_ENSEMBLE.rds")
rna.list <- SplitObject(stickleback.rna.comb, split.by = "orig.ident")
vst.list <- data.frame()
vst_one <- data.frame()
for (i in 1:length(rna.list)) {
    # cells_to_keep <- gsub("-1", "", rna.list[[i]]@assays[["RNA"]]@counts@Dimnames[[2]])
    cells_to_keep <- rna.list[[i]]@assays[["RNA"]]@counts@Dimnames[[2]]
    cells_to_keep <- subset(cells_to_keep, cells_to_keep %in% blood_cells_ids)
    rna.list[[i]] <- subset(rna.list[[i]], cells = cells_to_keep)
    rna.list[[i]] <- NormalizeData(rna.list[[i]], verbose = FALSE)
    rna.list[[i]] <- FindVariableFeatures(rna.list[[i]], selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
    tmp.frame <- data.frame(sample_vst = rna.list[[i]]@assays[["RNA"]]@meta.features[["vst.variance.standardized"]],
                            sample_ids = rna.list[[i]]@assays[["RNA"]]@counts@Dimnames[[1]])
    tmp.frame <- subset(tmp.frame, sample_ids %in% rna.list[[i]]@assays[["RNA"]]@var.features)
    tmp.frame <- tmp.frame[order(-tmp.frame$sample_vst),]
    tmp.frame <- mutate(tmp.frame, ID = row_number())
    tmp.frame$sample <- paste0("sample_", i)
    if(length(vst_one) == 0){
      
      vst_one <- tmp.frame
    }
    else{
      vst_one <- rbind(vst_one, tmp.frame)
    }
    tmp.frame$sample <- NULL
    
    colrenames <- c(paste0("sample_", i, "_variance_std"), paste0("sample_", i, "_ids"), "ID")
    colnames(tmp.frame) <- colrenames
    if(length(vst.list) == 0){
      vst.list <- tmp.frame
    }
    else{
      vst.list <- merge(vst.list, tmp.frame, by = "ID")
    }
}
# write.table(vst.list, file = "data/top_2000_HVGs_for_each_sample.tsv", sep = "\t")
# head(vst.list, 50)
 
hvg_union <- unique(c(vst.list$sample_1_ids,
               vst.list$sample_2_ids,
               vst.list$sample_3_ids,
               vst.list$sample_4_ids))

hvg_counts_sample_1 <- tibble::rownames_to_column(as.data.frame(rna.list[["stickleback.sample.1"]]@assays[["RNA"]]@counts[rownames(rna.list[["stickleback.sample.1"]]@assays[["RNA"]]@counts) %in% hvg_union,]))
hvg_counts_sample_2 <- tibble::rownames_to_column(as.data.frame(rna.list[["stickleback.sample.2"]]@assays[["RNA"]]@counts[rownames(rna.list[["stickleback.sample.2"]]@assays[["RNA"]]@counts) %in% hvg_union,]))
hvg_counts_sample_3 <- tibble::rownames_to_column(as.data.frame(rna.list[["stickleback.sample.3"]]@assays[["RNA"]]@counts[rownames(rna.list[["stickleback.sample.3"]]@assays[["RNA"]]@counts) %in% hvg_union,]))
hvg_counts_sample_4 <- tibble::rownames_to_column(as.data.frame(rna.list[["stickleback.sample.4"]]@assays[["RNA"]]@counts[rownames(rna.list[["stickleback.sample.4"]]@assays[["RNA"]]@counts) %in% hvg_union,]))

hvg_counts_saline <- merge(hvg_counts_sample_1, hvg_counts_sample_2, by="rowname")
hvg_counts_freshwater <- merge(hvg_counts_sample_3, hvg_counts_sample_4, by="rowname")

hvg_counts_all <- merge(hvg_counts_saline, hvg_counts_freshwater, by="rowname")
saline_right_boundary <- length(hvg_counts_saline) - 1
freshwater_left_boundary <- saline_right_boundary + 1 - 1
freshwater_right_boundary <- saline_right_boundary + length(hvg_counts_freshwater) - 1 - 1

output <- apply(hvg_counts_all[,2:freshwater_right_boundary], 1, function(row){
  # message(paste0("row = ", row))
  x <- row[1:saline_right_boundary]
  x <- x[x != 0]
  y <- row[freshwater_left_boundary:(freshwater_right_boundary)]
  y <- y[y != 0]
  if(length(x) > 10 & length(y) > 10){
    f.test.res <- var.test(x, y)
    return(data.frame(f_ratio = f.test.res$estimate, p_val = f.test.res$p.value))
  }else{
    return(data.frame(f_ratio = NA, p_val = NA))
  }
  
  # return(data.frame(test = sum(row[1:saline_right_boundary]), test.l.1 = length(row[1:saline_right_boundary]), test.l.2 = length(row[freshwater_left_boundary:freshwater_right_boundary])))
})
f.test.frame <- do.call(rbind,output)
rownames(f.test.frame) <- hvg_counts_all$rowname

```

```{r}
write_csv(tibble(HVGs = hvg_union), file = "plots/for_article/Suppl_Table_scRNA_HVGs_from_all_samples_unity_2000_from_each.csv")
 f.test.frame %>% 
  rownames_to_column(var = "gene") %>% 
  filter(is.finite(f_ratio), f_ratio > 0) %>% 
  write_csv(file = "plots/for_article/Suppl_Table_scRNA_F_test_table_with_HVGs_from_all_samples_cleaned.csv")
```


```{r}
 f.test.frame %>% 
  filter(is.finite(f_ratio), f_ratio > 0) %>% 
  ggplot(aes(x=f_ratio))+
  geom_histogram(binwidth = 0.25, fill = "#a2b1ba", color = "#899ba7") +
  geom_vline(aes(xintercept=median(f_ratio)), color="red", linetype="dashed", size=1) +
  coord_cartesian(xlim = c(0, 5)) +
  theme_classic(base_size = 20) +
  xlab("F-ratio") +
  ylab("Number of genes") -> p_f_ratio_blood

svg(filename = "plots/for_article/scRNA_F_ratio_histogram_for_blood_cluster_on_HVGs_from_all_samples.svg")
print(p_f_ratio_blood)
dev.off()
p_f_ratio_blood
```

### F-test histogram for blood cluster for all genes with no filtering

```{r}

saline_right_boundary <- table(rna.integrated.cluster.blood@meta.data[["orig.ident"]])[1] + table(rna.integrated.cluster.blood@meta.data[["orig.ident"]])[2]
freshwater_left_boundary <- saline_right_boundary + 1
freshwater_right_boundary <- saline_right_boundary + table(rna.integrated.cluster.blood@meta.data[["orig.ident"]])[3] + table(rna.integrated.cluster.blood@meta.data[["orig.ident"]])[4]

output <- apply(rna.integrated.cluster.blood@assays[["RNA"]]@counts, 1, function(row){
  x <- row[1:saline_right_boundary]
  x <- x[x != 0 & is.finite(x)]
  y <- row[freshwater_left_boundary:(freshwater_right_boundary)]
  y <- y[y != 0 & is.finite(y)]
  if(length(x) > 10 & length(y) > 10){
    saline_var <- var(x)
    freshwater_var <- var(y)
    f.test.res <- var.test(x, y)
    # if(is.finite(f.test.res$estimate) & f.test.res$estimate == 0){
    #   f.test.res$estimate = 1 # ========================================================== ARTIFICIALLY CHANGING ZEROES TO 1
    # }
    return(data.frame(f_ratio = f.test.res$estimate, p_val = f.test.res$p.value, saline_var = saline_var, freshwater_var = freshwater_var))
  }else{
    return(data.frame(f_ratio = NA, p_val = NA, saline_var = NA, freshwater_var = NA))
  }
})
f.test.frame_all_genes <- do.call(rbind,output)
# rownames(f.test.frame) <- hvg_counts_all$rowname

```





```{r}
 f.test.frame_all_genes %>% 
  filter(is.finite(f_ratio), f_ratio > 0) %>% 
  ggplot(aes(x=f_ratio))+
  geom_histogram(binwidth = 0.25, fill = "#a2b1ba", color = "#899ba7") +
  geom_vline(aes(xintercept=median(f_ratio)), color="red", linetype="dashed", size=1) +
  coord_cartesian(xlim = c(0, 5)) +
  theme_classic(base_size = 20) +
  xlab("F-ratio") +
  ylab("Number of genes") -> p_f_ratio_blood

svg(filename = "plots/for_article/scRNA_F_ratio_histogram_for_blood_cluster_all_genes_no_filtering.svg")
print(p_f_ratio_blood)
dev.off()
png(filename = "plots/for_article_png/scRNA_F_ratio_histogram_for_blood_cluster_all_genes_no_filtering.png", 
    res = 300, 
    units = "in",
    width = 7,
    height = 7)
print(p_f_ratio_blood)
dev.off()
p_f_ratio_blood
```

### F-test histogram for all cells for all genes with no filtering

```{r}

saline_right_boundary <- table(rna.integrated@meta.data[["orig.ident"]])[1] + table(rna.integrated@meta.data[["orig.ident"]])[2]
freshwater_left_boundary <- saline_right_boundary + 1
freshwater_right_boundary <- saline_right_boundary + table(rna.integrated@meta.data[["orig.ident"]])[3] + table(rna.integrated@meta.data[["orig.ident"]])[4]

output <- apply(rna.integrated@assays[["RNA"]]@counts, 1, function(row){
  x <- row[1:saline_right_boundary]
  x <- x[x != 0 & is.finite(x)]
  y <- row[freshwater_left_boundary:(freshwater_right_boundary)]
  y <- y[y != 0 & is.finite(y)]
  if(length(x) > 10 & length(y) > 10){
    saline_var <- var(x)
    freshwater_var <- var(y)
    f.test.res <- var.test(x, y)
    # if(is.finite(f.test.res$estimate) & f.test.res$estimate == 0){
    #   f.test.res$estimate = 1 # ========================================================== ARTIFICIALLY CHANGING ZEROES TO 1
    # }
    return(data.frame(f_ratio = f.test.res$estimate, p_val = f.test.res$p.value, saline_var = saline_var, freshwater_var = freshwater_var))
  }else{
    return(data.frame(f_ratio = NA, p_val = NA, saline_var = NA, freshwater_var = NA))
  }
})
f.test.frame_all_cells_genes <- do.call(rbind,output)

```



```{r}
 f.test.frame_all_cells_genes %>% 
  filter(is.finite(f_ratio), f_ratio > 0) %>% 
  ggplot(aes(x=f_ratio))+
  geom_histogram(binwidth = 0.25, fill = "#a2b1ba", color = "#899ba7") +
  geom_vline(aes(xintercept=median(f_ratio)), color="red", linetype="dashed", size=1) +
  coord_cartesian(xlim = c(0, 5)) +
  theme_classic(base_size = 20) +
  xlab("F-ratio") +
  ylab("Number of genes") -> p_f_ratio_all

svg(filename = "plots/for_article/Suppl_scRNA_F_ratio_histogram_for_all_cells_all_genes_no_filtering.svg")
print(p_f_ratio_all)
dev.off()
png(filename = "plots/for_article_png/Suppl_scRNA_F_ratio_histogram_for_all_cells_all_genes_no_filtering.png", 
    res = 300, 
    units = "in",
    width = 7,
    height = 7)
print(p_f_ratio_all)
dev.off()
p_f_ratio_all
```

# Venn digram of HVGs for all cells in scRNA-seq dataset

```{r subsetting_blood_for_each_sample_and_HVGs_all_cells}
load(file = "data/rna.integrated.filtered.after.clustering.obj.RData")

stickleback.rna.comb <- readRDS(file = "data/stickleback.rna.seurat.samples.PURE_ENSEMBLE.rds")
rna.list <- SplitObject(stickleback.rna.comb, split.by = "orig.ident")
vst.list_all_cells <- data.frame()
vst_one <- data.frame()
for (i in 1:length(rna.list)) {
    # cells_to_keep <- rna.list[[i]]@assays[["RNA"]]@counts@Dimnames[[2]]
    # cells_to_keep <- subset(cells_to_keep, cells_to_keep %in% blood_cells_ids)
    # rna.list[[i]] <- subset(rna.list[[i]], cells = cells_to_keep)
    rna.list[[i]] <- NormalizeData(rna.list[[i]], verbose = FALSE)
    rna.list[[i]] <- FindVariableFeatures(rna.list[[i]], selection.method = "vst", 
        nfeatures = 2000, verbose = FALSE)
    tmp.frame <- data.frame(sample_vst = rna.list[[i]]@assays[["RNA"]]@meta.features[["vst.variance.standardized"]],
                            sample_ids = rna.list[[i]]@assays[["RNA"]]@counts@Dimnames[[1]])
    tmp.frame <- subset(tmp.frame, sample_ids %in% rna.list[[i]]@assays[["RNA"]]@var.features)
    tmp.frame <- tmp.frame[order(-tmp.frame$sample_vst),]
    tmp.frame <- mutate(tmp.frame, ID = row_number())
    tmp.frame$sample <- paste0("sample_", i)
    if(length(vst_one) == 0){
      
      vst_one <- tmp.frame
    }
    else{
      vst_one <- rbind(vst_one, tmp.frame)
    }
    tmp.frame$sample <- NULL
    
    colrenames <- c(paste0("sample_", i, "_variance_std"), paste0("sample_", i, "_ids"), "ID")
    colnames(tmp.frame) <- colrenames
    if(length(vst.list_all_cells) == 0){
      vst.list_all_cells <- tmp.frame
    }
    else{
      vst.list_all_cells <- merge(vst.list_all_cells, tmp.frame, by = "ID")
    }
}
# write.table(vst.list, file = "data/top_2000_HVGs_for_each_sample_blood_only.tsv", sep = "\t")
# head(vst.list, 50)
```

### Venn diagrams for the overlapping HVGs for all samples (updated: only for blood)

```{r venn_diagrams_for_overlapping_HVGs}
require(venn)
    
# vst.list <- read.table(file = "data/top_2000_HVGs_for_each_sample_blood_only.tsv")

png(filename = "plots/for_article/Suppl_scRNA_venn_hvgs_per_sample_all_cells.png", res = 300, width = 1500, height =1200)
venn(list(`1_sal` = vst.list_all_cells$sample_1_ids,
          `2_sal` = vst.list_all_cells$sample_2_ids,
          `3_fr` = vst.list_all_cells$sample_3_ids,
          `4_fr` = vst.list_all_cells$sample_4_ids),
     zcolor = c("#4A8E9F", "#4A8E9F", "#A1887F", "#A1887F"), opacity = 0.8, ilcs = 1.2, sncs = 1, box = FALSE, ggplot = FALSE, plotsize = 60)
dev.off()
venn(list(`1_sal` = vst.list_all_cells$sample_1_ids,
          `2_sal` = vst.list_all_cells$sample_2_ids,
          `3_fr` = vst.list_all_cells$sample_3_ids,
          `4_fr` = vst.list_all_cells$sample_4_ids),
     zcolor = c("#4A8E9F", "#4A8E9F", "#A1887F", "#A1887F"), opacity = 0.8, ilcs = 1.2, sncs = 1, box = FALSE, ggplot = FALSE, plotsize = 60)
```

### GO enrichment with conversion to Zebrafish orthologs, background = union of all HVGs for all samples (ALL cells)

```{r stickleback_to_zebrafish_IDs_function}
conversion_table <- read.table(file = "data/stickleback_zebrafish_orthologs.tsv", sep = "\t", header = TRUE)
stickleback_to_zebrafish_IDs <- function(stickleback_ensembl_ids, conversion_table, id_type="zebrafish_ensembl"){
  conversion_table <- merge(data.frame(stickleback_ensembl = stickleback_ensembl_ids), conversion_table, by="stickleback_ensembl", sort=FALSE)
  zebrafish_ids <- conversion_table[,id_type]
  return(zebrafish_ids)
}
```

```{r GO_enrichment_watertype_specific_HVGs}

P_ADJUST_VALUE_CUTOFF = 0.2
# GO enrichment for HVGs common for freshwater and saline
saline_specific_HVGs <- stickleback_to_zebrafish_IDs(setdiff(intersect(vst.list_all_cells$sample_1_ids,
                                                                       vst.list_all_cells$sample_2_ids),
                                                             union(vst.list_all_cells$sample_3_ids,
                                                                   vst.list_all_cells$sample_4_ids)), conversion_table)
freshwater_specific_HVGs <- stickleback_to_zebrafish_IDs(setdiff(intersect(vst.list_all_cells$sample_3_ids,
                                                                           vst.list_all_cells$sample_4_ids),
                                                                 union(vst.list_all_cells$sample_1_ids,
                                                                       vst.list_all_cells$sample_2_ids)), conversion_table)
hvg_union_background <- stickleback_to_zebrafish_IDs(c(vst.list_all_cells$sample_1_ids,
                                                       vst.list_all_cells$sample_2_ids,
                                                       vst.list_all_cells$sample_3_ids,
                                                       vst.list_all_cells$sample_4_ids), conversion_table)
# write.table(hvg_union_background, file = "data/hvg_union_background_all_cells.tsv", sep = "\t")

# BiocManager::install("org.Dr.eg.db")
require(clusterProfiler)
saline_GO <- enrichGO(gene = saline_specific_HVGs,
                OrgDb         = "org.Dr.eg.db",
                keyType       = 'ENSEMBL',
                pAdjustMethod = "BH",
                pvalueCutoff = P_ADJUST_VALUE_CUTOFF,
                qvalueCutoff=1,
                universe = hvg_union_background)
freshwater_GO <- enrichGO(gene = freshwater_specific_HVGs,
                OrgDb         = "org.Dr.eg.db",
                keyType       = 'ENSEMBL',
                pAdjustMethod = "BH",
                pvalueCutoff = P_ADJUST_VALUE_CUTOFF,
                qvalueCutoff=1,
                universe = hvg_union_background)
```

   
```{r GO_enrichment_plots}
# Top of GO enrichment terms for oppositely changing genes between samples
require(enrichplot)

png(filename = "plots/for_article/Suppl_scRNA_GO_enrichment_waterspecific_HVGs_saline_ALL_CELLS.png", 
    res = 300, 
    units = "in",
    width = 9,
    height = 7)
dotplot(saline_GO, showCategory = 30,
        title = "GO terms, saline_GO")
dev.off()
png(filename = "plots/for_article/Suppl_scRNA_GO_enrichment_waterspecific_HVGs_freshwater_ALL_CELLS.png", 
    res = 300, 
    units = "in",
    width = 9,
    height = 7)
dotplot(freshwater_GO, showCategory = 30,
        title = "GO terms, freshwater_GO")
dev.off()

dotplot(saline_GO, showCategory = 30,
        title = "GO terms, saline_GO")
dotplot(freshwater_GO, showCategory = 30,
        title = "GO terms, freshwater_GO")
```

```{r to_entrez_for_kegg}
require(biomaRt)
ensembl <- useEnsembl(biomart = "ensembl")
ensembl <- useDataset(dataset = "drerio_gene_ensembl", mart = ensembl)

freshwater_specific_HVGs_entrez <- getBM(attributes=c("entrezgene_id"),
      filters=c('ensembl_gene_id'),
      values=freshwater_specific_HVGs,
      mart=ensembl)[["entrezgene_id"]]

saline_specific_HVGs_entrez <- getBM(attributes=c("entrezgene_id"),
      filters=c('ensembl_gene_id'),
      values=saline_specific_HVGs,
      mart=ensembl)[["entrezgene_id"]]

hvg_union_background_entrez <- getBM(attributes=c("entrezgene_id"),
      filters=c('ensembl_gene_id'),
      values=hvg_union_background,
      mart=ensembl)[["entrezgene_id"]]

```

### KEGG pathways for watertype specific HVGs for ALL CELLS

```{r kegg_watertype_specific_HVGs}
P_ADJUST_VALUE_CUTOFF_KEGG = 0.1

saline_kegg <- enrichKEGG(saline_specific_HVGs_entrez,
                          organism="dre",
                          universe=hvg_union_background_entrez,
                          pvalueCutoff=0.1,
                          pAdjustMethod="BH")
freshwater_kegg <- enrichKEGG(freshwater_specific_HVGs_entrez,
                          organism="dre",
                          universe=hvg_union_background_entrez,
                          pvalueCutoff=0.1,
                          pAdjustMethod="BH")
```


```{r kegg_plots_watertype_hvgs}
png(filename = "plots/for_article/Suppl_scRNA_KEGG_enrichment_waterspecific_HVGs_saline_ALL_CELLS.png", 
    res = 300, 
    units = "in",
    width = 9,
    height = 7)
dotplot(saline_kegg,showCategory = 30,title = "KEGG terms, saline")
dev.off()
# png(filename = "plots/for_article/Suppl_scRNA_KEGG_enrichment_waterspecific_HVGs_freshwater_ALL_CELLS.png", 
#     res = 300, 
#     units = "in",
#     width = 7,
#     height = 7)
# dotplot(freshwater_kegg,showCategory = 30,title = "KEGG terms, freshwater")
# dev.off()
dotplot(saline_kegg,showCategory = 30,title = "KEGG terms, saline")
# dotplot(freshwater_kegg,showCategory = 30,title = "KEGG terms, freshwater") # NO CATEGORIES DISCOVERED
```