---
title: "Plots for article"
output: html_notebook
---


```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/home/tema/work/skolkovo/fish_project/")
knitr::opts_chunk$set(echo=FALSE, message = FALSE)
```

```{r, results = 'hide'}
library(tidyverse)
library(Seurat)
```

```{r}
load(file = "data/rna.integrated.filtered.after.clustering.obj.RData")
rna.integrated.cluster.blood <- subset(rna.integrated, idents = 0)
rna.integrated.cluster.rest <- subset(rna.integrated, idents = c(1,2))
```


```{r}
png(filename = "plots/for_article/S_f_1_umap_all_by_samples.png", res = 300, width = 1800, height =900)
DimPlot(rna.integrated, reduction = "umap", group.by = "orig.ident")
dev.off()
```

```{r}
png(filename = "plots/for_article/f_1_umap_all_low_res.png", res = 300, width = 1500, height =1200)
DimPlot(rna.integrated, reduction = "umap", cols = c("#EF9A9A", "#80CBC4", "#9FA8DA"))
dev.off()
DimPlot(rna.integrated, reduction = "umap", cols = c("#EF9A9A", "#80CBC4", "#9FA8DA"))
```

```{r}
paste0("Number of cells in the whole data: ", dim(rna.integrated@assays[["RNA"]]@counts)[[2]])
```

```{r}
rna.integrated.cluster.rest <- FindNeighbors(rna.integrated.cluster.rest, dims = 1:15)
rna.integrated.cluster.rest  <- FindClusters(rna.integrated.cluster.rest, resolution = 0.3)
rna.integrated.cluster.rest <- RunUMAP(rna.integrated.cluster.rest, dims = 1:15)

rna.integrated.cluster.rest <- FindClusters(rna.integrated.cluster.rest, resolution = 0.1)
```


```{r}
new_cluster_labels <- c("leukocytes_1", "macrophages", "gill_integument", "low_quality", "neutrophils", "lymphocytes", "ionocytes", "epithelium", "leukocytes_2", "neurons")
names(new_cluster_labels) <- levels(rna.integrated.cluster.rest)
rna.integrated.cluster.rest_annotated <- RenameIdents(rna.integrated.cluster.rest, new_cluster_labels)
```

```{r}
#Set color pallete for Clusters
nclusters <- length(levels(Idents(rna.integrated.cluster.rest_annotated)))
# cluster_color_pallete <- RColorBrewer::brewer.pal(n = nclusters, name = 'Paired')
# cluster_color_pallete <- c("#c47d41",
# "#8462ca",
# "#67b64e",
# "#c861b1",
# "#bfad44",
# "#688ccd",
# "#ce4a3c",
# "#4bb193",
# "#c55d78",
# "#6a7e37")

# cluster_color_pallete <- c("#ff9017",
# "#00589a",
# "#c2a900",
# "#b793ff",
# "#00d365",
# "#d80087",
# "#5b2300",
# "#f7afee",
# "#f20037",
# "#43162e") #intense but distinct

cluster_color_pallete <- c("#4bb193",
"#bc6789",
"#6788cc",
"#6e8139",
"#bd7846",
"#c7a83f",
"#9a64ca",
"#d25141",
"#d34894",
"#6bb74d") # soft, earthy
```



```{r}
pdf(file = "plots/for_article/S_f_2_umap_rest_high_res_03_annotated.pdf")
DimPlot(rna.integrated.cluster.rest_annotated, reduction = "umap", cols = cluster_color_pallete)
dev.off()
DimPlot(rna.integrated.cluster.rest_annotated, reduction = "umap", cols = cluster_color_pallete)
```

```{r}
pdf(file = "plots/for_article/Suppl_scRNA_umap_rest_high_res_03_annotated_with_lables.pdf")
DimPlot(rna.integrated.cluster.rest_annotated, reduction = "umap", label = TRUE, repel = TRUE, cols = cluster_color_pallete)
dev.off()
DimPlot(rna.integrated.cluster.rest_annotated, reduction = "umap", label = TRUE, repel = TRUE, cols = cluster_color_pallete) 
```


```{r}
paste0("Number of cells in the non-blood data: ", dim(rna.integrated.cluster.rest_annotated@assays[["RNA"]]@counts)[[2]])
```

### Heatmap rest

```{r fig.height=42, fig.width=9}
integrated.cluster.rest.markers <- FindAllMarkers(rna.integrated.cluster.rest_annotated, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.15)
integrated.cluster.rest.markers %>%
    group_by(cluster) %>%
    top_n(n = 20, wt = avg_log2FC) -> top_by_logfc

integrated.cluster.rest.markers_RNA <- FindAllMarkers(rna.integrated.cluster.rest_annotated, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.15, assay = "RNA")
integrated.cluster.rest.markers_RNA %>%
    group_by(cluster) %>%
    # filter(avg_log2FC > 1) %>% 
    top_n(n = 20, wt = avg_log2FC) -> top_on_RNA

FindAllMarkers(rna.integrated.cluster.rest_annotated, 
               test.use = "MAST",
               only.pos = TRUE, 
               min.pct = 0.1, 
               logfc.threshold = 0.15, 
               assay = "RNA"
               ) %>%
    group_by(cluster) %>%
    top_n(n = 20, wt = avg_log2FC) -> top_MAST

manual_markers <- tibble(p_val = NA,
                        avg_log2FC = NA,
                        pct.1 = NA,
                        pct.2 = NA,
                        p_val_adj = NA,
                        cluster = factor(c("lymphocytes", "macrophages"), levels = levels(Idents(rna.integrated.cluster.rest_annotated))),
                        gene = c("lck", "mpeg1.2")
                        )

exclude_markers <- c("atp6v1g1", "UBL5", "ENSGACG00000009225", "acod1", "tsr3", "atp6v1e1b", "rpl22l1", "ENSGACG00000020069", "ddx18", "cnbpa", "sh2d1ab", "ENSGACG00000012769", "tnfsf14", "ENSGACG00000004750", "ENSGACG00000009171", "ENSGACG00000018643", "sla1a", "tagln2", "rpl19", "sftpbb", "nme2a", "nkl.4", "rpl29", "eif3i", "zgc:162944", "tubb2b", "ran", "rps13", "rps7", "fance", "rpl23a", "ENSGACG00000000326", "rplp2l", "ENSGACG00000000860", "ENSGACG00000007836", "rplp0", "rps26", "RPL19 (1 of many)", "rpl13a", "atp6v1g1", "UBL5", "ENSGACG00000009225", "ENSGACG00000007536", "ENSGACG00000001108", "cndp2", "fabp3", "zgc:152830", "mdh1aa", "ENSGACG00000000869", "rpl35a", "rps9", "rpl35", "ENSGACG00000013895", "HBE1 (1 of many)", "ENSGACG0000001390", "ENSGACG00000001394", "hbae5", "h2afva", "scinla", "ENSGACG00000020692", "ENSGACG00000001436", "capn2b", "chmp4c", "chmp2a", "gstr (1 of many)", "rhbg", "rdh12l", "si:dkey−87o1.2", "ENSGACG00000008205", "id4", "krt8", "ENSGACG00000006166", "ENSGACG00000006236", "ENSGACG00000004882", "ENSGACG00000017697", "nrarpa", "capn2b", "zgc:86896", "camk2g2", "cyp46a1.4", "plcg2", "plecb", "ENSGACG00000013902", "tpm4a", "sptbn2", "lmo7a", "pfkpb", "CKB", "itgb2", "tank", "scinlb", "nop56", "ENSGACG00000007176", "ENSGACG00000006275", "sat1b", "si:dkey−87o1.2", "HBE1 (1 of many).1")

rbind(top_by_logfc, top_on_RNA, top_MAST, manual_markers) %>% 
  transform(gene_cluster = paste0(gene, "_", cluster)) %>% 
  group_by(gene_cluster) %>%
  sample_n(1) %>%  # dropping duplicate markers
  arrange(cluster, -avg_log2FC) %>% 
  filter((gene %in% exclude_markers) == FALSE) -> top_markers_comb
# Number of cell with non-zero expression per cluster 
#     convert all non-zero values to 1
rest_count_matrix_binarized <- rna.integrated.cluster.rest_annotated@assays[["RNA"]]@counts # all non-zero peaks set to value 1
rest_count_matrix_binarized@x <- as.numeric(rep_len(1L, length(rest_count_matrix_binarized@x)))
#     sum column-wise by gene (resulting with a matrix of num_nonzero_peaks_in_gene vs cells)
expressing_cells_per_cluster_matrix <- t(rowsum(t(as.matrix(rest_count_matrix_binarized)), Idents(rna.integrated.cluster.rest_annotated)))

#     normalizing for the cluster size
norm_vector <- as.vector(table(Idents(rna.integrated.cluster.rest_annotated))/100)
expressing_cells_per_cluster_matrix <- t(t(expressing_cells_per_cluster_matrix)/norm_vector)

to_heatmap_rest <- expressing_cells_per_cluster_matrix[top_markers_comb$gene,]
rows_annot <- as.data.frame(top_markers_comb[,c("gene_cluster", "cluster")])
rownames(rows_annot)<- make.unique(rows_annot$gene_cluster)
rows_annot$gene_cluster <- NULL


annot_colors <- cluster_color_pallete
names(annot_colors) <- levels(Idents(rna.integrated.cluster.rest_annotated))
annot_colors <- list(cluster = annot_colors)

gene_names_heatmap <- rownames(to_heatmap_rest)
rownames(to_heatmap_rest) <- top_markers_comb$gene_cluster

breakpoints <- seq(0, 100, by = 10)
breakpoints[2] <- 0.1
cols <- colorRampPalette(RColorBrewer::brewer.pal(n = 11, name = "PuBu"))(length(breakpoints))


pheatmap::pheatmap(t(scale(t(to_heatmap_rest), center = FALSE)), 
                   cluster_cols = FALSE, 
                   cluster_rows = FALSE,  
                   labels_row = gene_names_heatmap,
                   # breaks = breakpoints,
                   # color = cols,
                   display_numbers = round(to_heatmap_rest, 1), # Side loading different values to display ! https://www.biostars.org/p/317989/
                   annotation_colors = annot_colors,
                   # scale = "row",
                   annotation_row = rows_annot
                   )-> heatmap_rest_avg_with_num
pdf(file = "plots/for_article/Suppl_scRNA_non_erythrocytes_heatmap_markers_10_clusters_percentage_of_expressing_cells_in_a_cluster_MORE_MARKERS_with_numbers.pdf", height = 42)
heatmap_rest_avg_with_num
dev.off()
```


```{r}
best_markers <- c("ENSGACG00000001123", "arpc1b", "coro1a", "ENSGACG00000014576", "ENSGACG00000009315", "swap70a", "thrap3a", "mpeg1.2", "ENSGACG00000018764", "krt99", "ca15b", "TYMP", "alas2", "tmem205", "gp1bb", "slc4a1a", "npsn (1 of many)", "fbp2", "ncf2", "mpx", "lck", "ENSGACG00000004719", "prf1.9", "ENSGACG00000006316", "atp1a1a.4", "slc4a1b", "atp1b1b", "cyc1", "krt15", "S100P", "ptgdsb.1", "ENSGACG00000011970", "ctsa", "lgmn", "ENSGACG00000003551", "cxcl19", "s100a1", "smpx", "camsap3", "tspan2a")
best_markers <- top_markers_comb[top_markers_comb$gene %in% best_markers,]

# Number of cell with non-zero expression per cluster 
#     convert all non-zero values to 1
rest_count_matrix_binarized <- rna.integrated.cluster.rest_annotated@assays[["RNA"]]@counts # all non-zero peaks set to value 1
rest_count_matrix_binarized@x <- as.numeric(rep_len(1L, length(rest_count_matrix_binarized@x)))
#     sum column-wise by gene (resulting with a matrix of num_nonzero_peaks_in_gene vs cells)
expressing_cells_per_cluster_matrix <- t(rowsum(t(as.matrix(rest_count_matrix_binarized)), Idents(rna.integrated.cluster.rest_annotated)))

#     normalizing for the cluster size
norm_vector <- as.vector(table(Idents(rna.integrated.cluster.rest_annotated))/100)
expressing_cells_per_cluster_matrix <- t(t(expressing_cells_per_cluster_matrix)/norm_vector)

to_heatmap_rest <- expressing_cells_per_cluster_matrix[best_markers$gene,]
rows_annot <- as.data.frame(best_markers[,c("gene_cluster", "cluster")])
rownames(rows_annot)<- make.unique(rows_annot$gene_cluster)
rows_annot$gene_cluster <- NULL


annot_colors <- cluster_color_pallete
names(annot_colors) <- levels(Idents(rna.integrated.cluster.rest_annotated))
annot_colors <- list(cluster = annot_colors)

gene_names_heatmap <- rownames(to_heatmap_rest)
rownames(to_heatmap_rest) <- best_markers$gene_cluster

pheatmap::pheatmap(t(scale(t(to_heatmap_rest), center = FALSE)), 
                   cluster_cols = FALSE, 
                   cluster_rows = FALSE,  
                   labels_row = gene_names_heatmap,
                   # breaks = breakpoints,
                   # color = cols,
                   # display_numbers = round(to_heatmap_rest, 1), # Side loading different values to display ! https://www.biostars.org/p/317989/
                   annotation_colors = annot_colors,
                   # scale = "row",
                   annotation_row = rows_annot
                   )-> heatmap_rest_avg_with_num
pdf(file = "plots/for_article/scRNA_non_erythrocytes_heatmap_10_clusters_percentage_of_expressing_cells_in_a_cluster_BEST_MARKERS.pdf", height = 7, width = 5)
heatmap_rest_avg_with_num
dev.off()
```



### Fig. 1d. Correlation of marine-freshwater expression divergence between bulk RNA-seq and averaged snRNA-seq datasets. 

```{r prepping_sc_DEGs}
# Loading DEGs generated by Seurat for the whole dataset saline vs freshwater
sc_DEGs <- read.csv(file = "data/DEGs.by.salinity.all.csv", header = TRUE)
id_match_table <- read.table(file = "data/10X_features.tsv", header = FALSE, sep = "\t", col.names = c("ensembl", "mixed", "type"))[,1:2]
sc_DEGs <- merge(sc_DEGs, id_match_table, by.x = "X", by.y = "mixed")
rownames(sc_DEGs) <- sc_DEGs$ensembl

sc_DEGs_filtered <- subset(sc_DEGs, abs(avg_logFC) > 0.25)
```

```{r calculation_bulk_DEGs}
bulk_DEGs <- as.data.frame(readxl::read_xlsx("data/S5C_table_RNAseq_DEG_MMFF.xlsx"))
rownames(bulk_DEGs) <- bulk_DEGs$...1
bulk_DEGs_common <- bulk_DEGs[sc_DEGs$ensembl, ]

bulk_DEGs_common_filtered <- bulk_DEGs[sc_DEGs_filtered$ensembl, ]
```

```{r plotting_corr_bulk_vs_sc}
require(ggpubr)
#combining DEG tablesf
combined_DEGs <- merge(sc_DEGs[,c("avg_logFC", "p_val", "ensembl")], bulk_DEGs_common[,c("logFC", "PValue", "...1")], by.x = "ensembl", by.y = "...1")
colnames(combined_DEGs) <- c("ensembl", "single_cell_logFC", "single_cell_pval", "bulk_logFC", "bulk_pval")
rownames(combined_DEGs) <- combined_DEGs$ensembl
write.table(combined_DEGs, file = "data/DEGs_overlapping_sc_and_bulk.tsv", sep = "\t")
combined_DEGs$density <- densCols(combined_DEGs$single_cell_logFC, combined_DEGs$bulk_logFC, colramp = colorRampPalette(rev(rainbow(10, end = 4/6))), nbin = 1024)



p_corr_bulk_vs_sc <- ggplot(combined_DEGs) +
      geom_point(aes(single_cell_logFC, bulk_logFC, col = density), size = 1) +
      stat_cor(aes(single_cell_logFC, bulk_logFC), method = "pearson", label.x = 0.03, label.y = -2, size=6, color="#696969")+
      geom_smooth(aes(single_cell_logFC, bulk_logFC), method=lm, se=FALSE, fullrange=TRUE, linetype="dashed", color="dark grey") +
      coord_cartesian(xlim = c(-0.15, 0.15), ylim = c(-2.5, 2.25)) +
      # xlim(-0.25, 0.25) +
      # ylim(-2.5, 2.25) +
      scale_color_identity() +
      theme_classic(base_size = 20) +
      theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) +
      xlab("logFC, scRNA-seq") +
      ylab("logFC, RNA-seq")

p_corr_bulk_vs_sc
svg(filename = "plots/for_article/f_1d_scatter_scRNA_vs_bulk_with_density.svg", width = 7, height = 7)
print(p_corr_bulk_vs_sc)
dev.off()
```



```{r mean_of_all_sc_counts}
library(tidyverse)
rna.integrated.saline <- subset(rna.integrated, orig.ident %in% c("stickleback.sample.1", "stickleback.sample.2"))
rna.integrated.fresh <- subset(rna.integrated, orig.ident %in% c("stickleback.sample.3", "stickleback.sample.4"))

saline_sc_counts <- as_tibble(rna.integrated.saline@assays[["RNA"]]@data)
saline_sc_counts %>% apply(.,1,mean) -> saline_sc_counts_mean
saline_sc_counts_mean <- tibble(gene = rna.integrated@assays[["RNA"]]@data@Dimnames[[1]], mean_expression = saline_sc_counts_mean)
cellranger_gene_table <- read.table(file = "data/10X_features.tsv", sep = "\t")
saline_sc_counts_mean <- merge(saline_sc_counts_mean, cellranger_gene_table[,1:2], by.x = "gene", by.y = "V2")
colnames(saline_sc_counts_mean) <- c("gene", "sc_mean_expression", "ensembl_id")

freshwater_sc_counts <- as_tibble(rna.integrated.fresh@assays[["RNA"]]@data)
freshwater_sc_counts %>% apply(.,1,mean) -> freshwater_sc_counts_mean
freshwater_sc_counts_mean <- tibble(gene = rna.integrated@assays[["RNA"]]@data@Dimnames[[1]], mean_expression = freshwater_sc_counts_mean)
cellranger_gene_table <- read.table(file = "data/10X_features.tsv", sep = "\t")
freshwater_sc_counts_mean <- merge(freshwater_sc_counts_mean, cellranger_gene_table[,1:2], by.x = "gene", by.y = "V2")
colnames(freshwater_sc_counts_mean) <- c("gene", "sc_mean_expression", "ensembl_id")
```

```{r}
# Function leaves randomly on of the duplicated rows
RemoveDups <- function(df, column) {  
  inds = sample(1:nrow(df))  
  df   = df[inds, ]

  dups = duplicated(df[, column])
  df   = df[!dups, ]
  inds = inds[!dups]

  df[sort(inds, index=T)$ix, ]
}
```


```{r normalized_bulk_counts}
library(edgeR)
library(magrittr)

stickleback_gene_lengths <- read.table(file = "data/stickleback_gene_lengths.txt", sep = "\t", header = TRUE)
stickleback_gene_lengths %<>% 
  rename(transcript_length = Transcript.length..including.UTRs.and.CDS.)

bulk_counts <- readxl::read_excel("data/msx156_Supp/S5A_table_RNAseq_rawcounts.xlsx")

bulk_counts_freshwater <-  
  select(bulk_counts, contains("FF") | contains("...1")) %>% rename(ensembl_id = ...1)

 merge(bulk_counts_freshwater,stickleback_gene_lengths, by.x="ensembl_id", by.y="Gene.stable.ID") %>% select(ensembl_id, transcript_length) -> gene_lengths
gene_lengths %<>% RemoveDups(., "ensembl_id")

y_freshwater <- DGEList(counts=select(bulk_counts_freshwater, !contains("id")),genes=gene_lengths)
y_freshwater <- calcNormFactors(y_freshwater)
bulk_counts_RPKM_freshwater <- as_tibble(rpkm(y_freshwater), gene.length = y_freshwater[["genes"]][["transcript_length"]])
bulk_counts_RPKM_freshwater_mean <- as_tibble(apply(bulk_counts_RPKM_freshwater, 1, mean))
  bulk_counts_RPKM_freshwater_mean$gene <- y_freshwater[["genes"]][["ensembl_id"]]
bulk_counts_RPKM_freshwater_mean %<>%
    rename(bulk_mean_expression = value)


bulk_counts_saline <-  
  select(bulk_counts, contains("FF") | contains("...1")) %>% rename(ensembl_id = ...1)

 merge(bulk_counts_saline,stickleback_gene_lengths, by.x="ensembl_id", by.y="Gene.stable.ID") %>% select(ensembl_id, transcript_length) -> gene_lengths
gene_lengths %<>% RemoveDups(., "ensembl_id")

y_saline <- DGEList(counts=select(bulk_counts_saline, !contains("id")),genes=gene_lengths)
y_saline <- calcNormFactors(y_saline)
bulk_counts_RPKM_saline <- as_tibble(rpkm(y_saline), gene.length = y_saline[["genes"]][["transcript_length"]])
bulk_counts_RPKM_saline_mean <- as_tibble(apply(bulk_counts_RPKM_saline, 1, mean))
  bulk_counts_RPKM_saline_mean$gene <- y_saline[["genes"]][["ensembl_id"]]
bulk_counts_RPKM_saline_mean %<>%
    rename(bulk_mean_expression = value)
```

```{r plotting_sc_vs_bulk_mean_counts_freshwater}

combined_counts_freshwater <- merge(freshwater_sc_counts_mean[c("sc_mean_expression", "ensembl_id")], bulk_counts_RPKM_freshwater_mean, by.x = "ensembl_id", by.y = "gene") 
combined_counts_freshwater %<>% mutate(sc_mean_expression = sc_mean_expression * (mean(bulk_mean_expression) / mean(sc_mean_expression) ))

write.table(combined_counts_freshwater, file = "data/combined_mean_expression_freshwater_sc_and_bulk.tsv", sep = "\t")
# combined_counts_freshwater$density <- densCols(combined_counts_freshwater$sc_mean_expression,
#                                                combined_counts_freshwater$bulk_mean_expression,
#                                                colramp = colorRampPalette(rev(rainbow(10, end = 4/6))),
#                                                nbin = 1024)



p_corr_bulk_vs_sc_freshwater <- ggplot(combined_counts_freshwater) +
      geom_point(aes(sc_mean_expression, bulk_mean_expression), size = 1) +
      # geom_point(aes(sc_mean_expression, bulk_mean_expression, col = density), size = 1) +
      stat_cor(aes(sc_mean_expression, bulk_mean_expression), method = "pearson", label.x = 50, label.y = 400, size=6, color="#696969")+
      coord_cartesian(xlim = c(0, 500), ylim = c(0, 500)) +
      # xlim(-0.25, 0.25) +
      # ylim(-2.5, 2.25) +
      scale_color_identity() +
      theme_classic(base_size = 20) +
      theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) +
      xlab("Expression, scRNA-seq") +
      ylab("Expression, RNA-seq")

p_corr_bulk_vs_sc_freshwater
svg(filename = "plots/for_article/S_f_3a_mean_experssion_freshwater_scatter_scRNA_vs_bulk_with_density.svg")
print(p_corr_bulk_vs_sc_freshwater)
dev.off()
```

```{r bulk_vs_sc_log_scale}
p_corr_bulk_vs_sc_freshwater <- ggplot(combined_counts_freshwater) +
      geom_point(aes(log(sc_mean_expression), log(bulk_mean_expression)), size = 1) +
      # geom_point(aes(sc_mean_expression, bullog()ean_expression, col = density), size = 1) +
      stat_cor(aes(log(sc_mean_expression), log(bulk_mean_expression)), method = "pearson", label.x = 4, label.y = 10, size=6, color="#696969")+
      # coord_cartesian(xlim = c(0, 500), ylim = c(0, 500)) +
      # xlim(-0.25, 0.25) + 
      # ylim(-2.5, 2.25) +
      scale_color_identity() +
      theme_classic(base_size = 20) +
      theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) +
      xlab("Expression, scRNA-seq") +
      ylab("Expression, RNA-seq")

p_corr_bulk_vs_sc_freshwater
```


```{r plotting_sc_vs_bulk_mean_counts_saline}

combined_counts_saline <- merge(saline_sc_counts_mean[c("sc_mean_expression", "ensembl_id")], bulk_counts_RPKM_saline_mean, by.x = "ensembl_id", by.y = "gene") 
combined_counts_saline %<>% mutate(sc_mean_expression = sc_mean_expression * (mean(bulk_mean_expression) / mean(sc_mean_expression) ))

write.table(combined_counts_saline, file = "data/combined_mean_expression_saline_sc_and_bulk.tsv", sep = "\t")
# combined_counts_saline$density <- densCols(combined_counts_saline$sc_mean_expression,
#                                                combined_counts_saline$bulk_mean_expression,
#                                                colramp = colorRampPalette(rev(rainbow(10, end = 4/6))),
#                                                nbin = 1024)



p_corr_bulk_vs_sc_saline <- ggplot(combined_counts_saline) +
      geom_point(aes(sc_mean_expression, bulk_mean_expression), size = 1) +
      # geom_point(aes(sc_mean_expression, bulk_mean_expression, col = density), size = 1) +
      stat_cor(aes(sc_mean_expression, bulk_mean_expression), method = "pearson", label.x = 50, label.y = 400, size=6, color="#696969")+
      coord_cartesian(xlim = c(0, 500), ylim = c(0, 500)) +
      # xlim(-0.25, 0.25) +
      # ylim(-2.5, 2.25) +
      scale_color_identity() +
      theme_classic(base_size = 20) +
      theme(panel.border = element_rect(colour = "black", fill=NA, size=1)) +
      xlab("Expression, scRNA-seq") +
      ylab("Expression, RNA-seq")

p_corr_bulk_vs_sc_saline
svg(filename = "plots/for_article/S_f_3b_mean_experssion_saline_scatter_scRNA_vs_bulk_with_density.svg")
print(p_corr_bulk_vs_sc_saline)
dev.off()
```

Insignificant, should not be included in the paper.

```{r number_of_cells_in_data}
# paste0("Number of cells in the whole dataset: ", dim(rna.integrated.cluster.rest@assays[["RNA"]]@counts)[[2]],
       # "\n Percent ")
```

### F-test histogram for blood cluster
```{r}
f.test.filtered.blood <- read.table(file = "data/f.test.filtered.blood.tsv", sep = "\t", header = TRUE)
p_f_ratio_blood <- ggplot(f.test.filtered.blood, aes(x=f_ratio))+
  geom_histogram(binwidth = 0.25) +
  geom_vline(aes(xintercept=median(f_ratio)), color="blue", linetype="dashed", size=1) +
  coord_cartesian(xlim = c(0, 5)) +
  theme_classic(base_size = 20) +
  xlab("F-ratio") +
  ylab("Number of genes")

svg(filename = "plots/for_article/f_1f_F_ratio_histogram_for_blood_cluster.svg")
print(p_f_ratio_blood)
dev.off()
p_f_ratio_blood
```


```{r dist_hist_blood_cluster_btw_phenotype_variability}
load(file = "data/rna.integrated.filtered.after.clustering.obj.RData")
rna.integrated.cluster.blood <- subset(rna.integrated, idents = 0)

rna.integrated.cluster.blood <- FindNeighbors(rna.integrated.cluster.blood, dims = 1:15)
rna.integrated.cluster.blood  <- FindClusters(rna.integrated.cluster.blood, resolution = 0.3)
rna.integrated.cluster.blood <- RunPCA(rna.integrated.cluster.blood)

set.seed(42)
pca_raw_blood <- as.data.frame(rna.integrated.cluster.blood@reductions[["pca"]]@cell.embeddings)
pca_raw_blood$sample <- rna.integrated.cluster.blood@meta.data[["orig.ident"]]
pca_raw_blood <- pca_raw_blood[sample(nrow(pca_raw_blood)), c("PC_1", "PC_2","PC_3","PC_4","PC_5","PC_6","PC_7","PC_8","PC_9","PC_10","sample")]

pca_saline_1_half <- subset(pca_raw_blood, sample == "stickleback.sample.1")[,1:10]
pca_saline_2_half <- subset(pca_raw_blood, sample == "stickleback.sample.2")[,1:10]
dist_matrix_salineVsaline <- fields::rdist(pca_saline_1_half, pca_saline_2_half)

pca_freshwater_1_half <- subset(pca_raw_blood, sample == "stickleback.sample.3")[,1:10]
pca_freshwater_2_half <- subset(pca_raw_blood, sample == "stickleback.sample.4")[,1:10]

dist_matrix_freshwaterVfreshwater <- fields::rdist(pca_freshwater_1_half, pca_freshwater_2_half)

frame_for_hist <- data.frame(dist = c(as.vector(dist_matrix_salineVsaline), as.vector(dist_matrix_freshwaterVfreshwater)),
                             pair = c(rep("salineVsaline", length(dist_matrix_salineVsaline)),
                                      rep("freshwaterVfreshwater", length(dist_matrix_freshwaterVfreshwater))))

mu_btw <- plyr::ddply(frame_for_hist, "pair", summarise, grp.mean=mean(dist))

small_frame_for_hist <- frame_for_hist[sample(nrow(frame_for_hist), 100000),]

p <- ggplot(small_frame_for_hist, aes(x=dist, fill=pair, color=pair)) +
  geom_histogram(aes(y = ..ncount..), position="identity", alpha=0.5, binwidth = 0.15) + 
  geom_vline(data=mu_btw, aes(xintercept=grp.mean, color=pair),
             linetype="dashed")+
  coord_cartesian(xlim = c(0, 15)) +
  theme_classic(base_size = 20) +
  xlab("Distance between pair of cells") +
  ylab("Number cell pairs")
p
svg(filename = "plots/for_article/f_1e_histogram_of_distance_btw_cells_within_saline_and_freshwater_only_blood_equalized_heights.svg", width = 11, height = 6)
print(p)
dev.off()
```


```{r pairwise_histogram_with_downsampling, fig.width=11}
load(file = "data/rna.integrated.filtered.after.clustering.obj.RData")
rna.integrated.cluster.blood <- subset(rna.integrated, idents = 0)

rna.integrated.cluster.blood <- FindNeighbors(rna.integrated.cluster.blood, dims = 1:15)
rna.integrated.cluster.blood  <- FindClusters(rna.integrated.cluster.blood, resolution = 0.3)
rna.integrated.cluster.blood <- RunPCA(rna.integrated.cluster.blood)

set.seed(42)
pca_raw_blood <- as.data.frame(rna.integrated.cluster.blood@reductions[["pca"]]@cell.embeddings)
pca_raw_blood$sample <- rna.integrated.cluster.blood@meta.data[["orig.ident"]]
pca_raw_blood <- pca_raw_blood[sample(nrow(pca_raw_blood)), c("PC_1", "PC_2","PC_3","PC_4","PC_5","PC_6","PC_7","PC_8","PC_9","PC_10","sample")]

pca_saline_1_half <- subset(pca_raw_blood, sample == "stickleback.sample.1")[,1:10]
pca_saline_2_half <- subset(pca_raw_blood, sample == "stickleback.sample.2")[,1:10]
pca_saline_2_half <- pca_saline_2_half[sample(nrow(pca_saline_2_half), nrow(pca_saline_1_half)),] # ========== DOWNSAMPLING CELL COUNT TO BALANCE SAMPLING FROM BOTH FISH
 
dist_matrix_salineVsaline <- fields::rdist(pca_saline_1_half, pca_saline_2_half)

pca_freshwater_1_half <- subset(pca_raw_blood, sample == "stickleback.sample.3")[,1:10]
pca_freshwater_1_half <- pca_freshwater_1_half[sample(nrow(pca_freshwater_1_half), nrow(pca_saline_1_half)),]
pca_freshwater_2_half <- subset(pca_raw_blood, sample == "stickleback.sample.4")[,1:10]
pca_freshwater_2_half <- pca_freshwater_2_half[sample(nrow(pca_freshwater_2_half), nrow(pca_freshwater_1_half)),]

dist_matrix_freshwaterVfreshwater <- fields::rdist(pca_freshwater_1_half, pca_freshwater_2_half)

frame_for_hist <- data.frame(dist = c(as.vector(dist_matrix_salineVsaline), as.vector(dist_matrix_freshwaterVfreshwater)),
                             pair = c(rep("salineVsaline", length(dist_matrix_salineVsaline)),
                                      rep("freshwaterVfreshwater", length(dist_matrix_freshwaterVfreshwater))))

mu_btw <- plyr::ddply(frame_for_hist, "pair", summarise, grp.mean=mean(dist))

small_frame_for_hist <- frame_for_hist[sample(nrow(frame_for_hist), 100000),]

p <- ggplot(small_frame_for_hist, aes(x=dist, fill=pair, color=pair)) +
  geom_histogram(aes(y = ..ncount..), position="identity", alpha=0.5, binwidth = 0.15) + 
  geom_vline(data=mu_btw, aes(xintercept=grp.mean, color=pair),
             linetype="dashed")+
  # coord_cartesian(xlim = c(0, 15)) +
  theme_classic(base_size = 20) +
  xlab("Distance between pair of cells") +
  ylab("Number cell pairs")
p
svg(filename = "plots/for_article/Suppl_histogram_of_distance_btw_cells_within_saline_and_freshwater_only_blood_equalized_heights_WITH_DOWNSAMPLING.svg", width = 11, height = 6)
print(p)
dev.off()
```

```{r means_table_blood_cluster_btw_phenotype_variability}
mu_btw
```

# F-ratio without filtering

```{r}
# f.test.filtered.blood <- read.table(file = "data/f.test.filtered.blood.tsv", sep = "\t", header = TRUE)
# MISSING IMPORT OF UNFILTERED TABLE
f.test.frame <- subset(f.test.frame, is.finite(f.test.frame$f_ratio))
p_f_ratio_blood <- ggplot(f.test.frame, aes(x=f_ratio))+
  geom_histogram(binwidth = 0.25) +
  geom_vline(aes(xintercept=median(f_ratio)), color="blue", linetype="dashed", size=1) +
  coord_cartesian(xlim = c(0, 5)) +
  theme_classic(base_size = 20) +
  xlab("F-ratio") +
  ylab("Number of genes")

# svg(filename = "plots/for_article/f_1f_F_ratio_histogram_for_blood_cluster.svg")
# print(p_f_ratio_blood)
# dev.off()
p_f_ratio_blood
```



