---
title: "GO analysis of top genes differential in variance according to two methods for saline/freshwater"
output: html_notebook
---

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "/home/tema/work/skolkovo/fish_project/")
knitr::opts_chunk$set(echo=FALSE, message = FALSE)
```

```{r, results = 'hide'}
library(tidyverse)
# library(Signac)
# library(Seurat)
# library(GenomeInfoDb)
# library(patchwork)
set.seed(42)
```

```{r}
#TODO:
#  - sort two lists for genes with differential varience by number of peaks and by the peak height by f-value 
#  - ?convert list based on peak heights to a list of genes with the most variable peaks in them?
#  - intersect top and bottom n (probably 100) genes from both sorted lists
#  - do GO analysis on the intersected subsets of genes
```

```{r importing_lists_with_differential_varience_calculated_two_ways}
var_table_peaks_binarized <- read_csv(file = "data/scATAC/saline_vs_freshwater_varience_table_on_balanced_matrix_with_number_of_peaks_per_gene_filtered.csv")
var_table_peaks_height <- read_csv(file = "data/scATAC/saline_vs_freshwater_varience_table_on_balanced_matrix_based_on_peaks_height_filtered.csv")
```

```{r sorting_and_taking_top_and_bottom}
var_table_peaks_binarized <- arrange(var_table_peaks_binarized, -f_ratio)
var_table_peaks_height <- arrange(var_table_peaks_height, -f_ratio)
```

```{r stickleback_to_zebrafish_IDs_function}
conversion_table <- read.table(file = "data/stickleback_zebrafish_orthologs.tsv", sep = "\t", header = TRUE)
stickleback_to_zebrafish_IDs <- function(stickleback_ensembl_ids, conversion_table, id_type="zebrafish_ensembl"){
  conversion_table <- merge(data.frame(stickleback_ensembl = stickleback_ensembl_ids), conversion_table, by="stickleback_ensembl", sort=FALSE)
  zebrafish_ids <- conversion_table[,id_type]
  return(zebrafish_ids)
}
```

```{r intersecting_tops_and_bottoms}
freshwater_top_threshold <- 2
# choosing arbitrarily to include peaks and genes with high varience difference (f_ratio) > 2, more varience in freshwater
var_table_peaks_binarized %>% filter(f_ratio > freshwater_top_threshold) -> top_binarized
var_table_peaks_height %>% filter(f_ratio > freshwater_top_threshold) -> top_height
top_genes_freshwater <- stickleback_to_zebrafish_IDs(intersect(top_binarized$gene, top_height$gene_id), conversion_table)

saline_top_threshold <- 0.7
var_table_peaks_binarized %>% filter(f_ratio < saline_top_threshold) -> bottom_binarized
var_table_peaks_height %>% filter(f_ratio < saline_top_threshold) -> bottom_height
top_genes_saline <- stickleback_to_zebrafish_IDs(intersect(bottom_binarized$gene, bottom_height$gene_id), conversion_table)
```

```{r GO_analysis_of_the_subsets}
require(clusterProfiler)
GO_p_value_threshold <- 0.1
freshwater_top_GO <- enrichGO(gene = top_genes_freshwater,
                OrgDb         = "org.Dr.eg.db",
                keyType       = 'ENSEMBL',
                pAdjustMethod = "BH",
                pvalueCutoff = GO_p_value_threshold,
                qvalueCutoff=1)

saline_top_GO <- enrichGO(gene = top_genes_saline,
                OrgDb         = "org.Dr.eg.db",
                keyType       = 'ENSEMBL',
                pAdjustMethod = "BH",
                pvalueCutoff = GO_p_value_threshold,
                qvalueCutoff=1)
```

Two lists of differentially variable genes were generated with two methods: taking the peak heights per gene in each cell and taking the number of non-zero valued peaks in a gene in each cell. Then the lists were filtered for f_ratio > `r freshwater_top_threshold` for top genes with high varience in freshwater and for f_ratio < `r saline_top_threshold` for high varience genes in saline. Then the filtered top subsets of the lists were intersected. The resulting 2 sets of genes were converted to Zebrafish ortologs and given as an input to GO analysis with filtering for p-value < `r GO_p_value_threshold`. Top GO terms are plotted below.

```{r GO_enrichment_plots, fig.width=12}
require(enrichplot)
dotplot(saline_top_GO, showCategory = 30,
        title = "top 30 GO terms, saline_top_GO")

```



```{r fig.width=12}
dotplot(freshwater_top_GO, showCategory = 30,
        title = "top 30 GO terms, freshwater_top_GO")
```


